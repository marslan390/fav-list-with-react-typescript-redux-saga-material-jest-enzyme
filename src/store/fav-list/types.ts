export interface FavItem {
  id?: number;
  name: string;
  author: string;
  rating: number;
}

export interface FavListState {
  items: FavItem[];
  isFetching: boolean;
  isRating: boolean;
}

export const DEFAULT_SORT_PROP = "rating";

export const GET_FAV_LIST = "GET_FAV_LIST";
export const RECEIVE_FAV_LIST = "RECEIVE_FAV_LIST";
export const RECEIVE_FAV_LIST_ERROR = "RECEIVE_FAV_LIST_ERROR";
export const RATE_FAV_ITEM = "RATE_FAV_ITEM";
export const TOGGLE_RANDOM_FAV_RATING = "TOGGLE_RANDOM_FAV_RATING";

interface GetFavListAction {
  type: typeof GET_FAV_LIST;
}

interface ReceiveFavListAction {
  type: typeof RECEIVE_FAV_LIST;
  payload: FavItem[];
}

interface ReceiveFavListErrorAction {
  type: typeof RECEIVE_FAV_LIST_ERROR;
  payload: FavItem[];
}

interface RateFavItemAction {
  type: typeof RATE_FAV_ITEM;
  payload: FavItem;
}

interface ToggleRandomRatingAction {
  type: typeof TOGGLE_RANDOM_FAV_RATING;
}

export type FavListActions =
  | GetFavListAction
  | ReceiveFavListAction
  | ReceiveFavListErrorAction
  | RateFavItemAction
  | ToggleRandomRatingAction;
