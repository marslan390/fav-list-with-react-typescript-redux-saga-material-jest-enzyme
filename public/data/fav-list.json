[
  {
    "id": 1,
    "name": "The Hunger Games",
    "author": "Suzanne Collins",
    "rating": 4.33
  },
  {
    "id": 2,
    "name": "Harry Potter",
    "author": "J.K. Rowling",
    "rating": 4.49
  },
  {
    "id": 3,
    "name": "To Kill a Mockingbird",
    "author": "Harper Lee",
    "rating": 4.27
  },
  {
    "id": 4,
    "name": "Pride and Prejudice",
    "author": "Jane Austen",
    "rating": 4.26
  },
  {
    "id": 5,
    "name": "Twilight",
    "author": "Stephenie Meyer",
    "rating": 4.33
  },
  {
    "id": 6,
    "name": "The Book Thief",
    "author": "Markus Zusak",
    "rating": 4.37
  },
  {
    "id": 7,
    "name": "Animal Farm",
    "author": "George Orwell",
    "rating": 3.39
  },
  {
    "id": 8,
    "name": "The Chronicles of Narnia",
    "author": "C.S. Lewis",
    "rating": 4.26
  },
  {
    "id": 9,
    "name": "The Hobbit and The Lord of the Rings",
    "author": "J.R.R. Tolkien",
    "rating": 4.59
  },
  {
    "id": 10,
    "name": "Gone with the Wind",
    "author": "Margaret Mitchell",
    "rating": 4.29
  },
  {
    "id": 11,
    "name": "The Fault in Our Stars",
    "author": "John Green",
    "rating": 4.22
  },
  {
    "id": 12,
    "name": "The Giving Tree",
    "author": "Shel Silverstein",
    "rating": 4.37
  },
  {
    "id": 13,
    "name": "Wuthering Heights",
    "author": "Emily BrontÃ«",
    "rating": 3.85
  },
  {
    "id": 14,
    "name": "The Da Vinci Code",
    "author": "Dan Brown",
    "rating": 3.84
  },
  {
    "id": 15,
    "name": "Memoirs of a Geisha",
    "author": "Arthur Golden",
    "rating": 5
  }
]
