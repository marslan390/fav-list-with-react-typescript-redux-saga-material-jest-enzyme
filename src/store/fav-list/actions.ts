import {
  FavItem,
  FavListActions,
  GET_FAV_LIST,
  RATE_FAV_ITEM,
  RECEIVE_FAV_LIST,
  TOGGLE_RANDOM_FAV_RATING,
} from "./types";

export const getFavList = (): FavListActions => ({
  type: GET_FAV_LIST,
});

export const receiveFavList = (favItems: FavItem[]): FavListActions => ({
  type: RECEIVE_FAV_LIST,
  payload: favItems,
});

export const receiveFavListError = (errors: any): FavListActions => ({
  type: RECEIVE_FAV_LIST,
  payload: errors,
});

export const rateFavItem = (favItem: FavItem): FavListActions => ({
  type: RATE_FAV_ITEM,
  payload: favItem,
});

export const toggleRandomFavRating = (): FavListActions => ({
  type: TOGGLE_RANDOM_FAV_RATING,
});
