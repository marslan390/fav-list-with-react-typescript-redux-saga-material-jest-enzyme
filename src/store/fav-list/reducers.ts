import {
  DEFAULT_SORT_PROP,
  FavItem,
  FavListActions,
  FavListState,
  GET_FAV_LIST,
  RATE_FAV_ITEM,
  RECEIVE_FAV_LIST,
  RECEIVE_FAV_LIST_ERROR,
  TOGGLE_RANDOM_FAV_RATING
} from "./types";
import { sortDesc } from "../../utils/array";

export const favListInitialState: FavListState = {
  items: [],
  isFetching: false,
  isRating: false
};

export const favListReducer = (
  state: FavListState = favListInitialState,
  action: FavListActions
): FavListState => {
  switch (action.type) {
    case GET_FAV_LIST: {
      return {
        ...state,
        isFetching: true
      };
    }
    case RECEIVE_FAV_LIST_ERROR: {
      return {
        ...state,
        isFetching: false
      };
    }
    case RECEIVE_FAV_LIST: {
      return {
        ...state,
        items: getSortedFavList(action.payload),
        isFetching: false
      };
    }
    case RATE_FAV_ITEM: {
      return {
        ...state,
        items: getFavListAfterRate([...state.items], action.payload)
      };
    }
    case TOGGLE_RANDOM_FAV_RATING: {
      return {
        ...state,
        isRating: !state.isRating
      };
    }
    default:
      return state;
  }
};

const getSortedFavList = (items: FavItem[]): FavItem[] => {
  return sortDesc([...items], DEFAULT_SORT_PROP);
};

const getFavListAfterRate = (items: FavItem[], favItem: FavItem): FavItem[] => {
  const _favList = [...items];
  const index = items.findIndex(item => item.id === favItem.id);
  if (index !== -1) {
    _favList[index] = favItem;
  }
  return getSortedFavList(_favList);
};
