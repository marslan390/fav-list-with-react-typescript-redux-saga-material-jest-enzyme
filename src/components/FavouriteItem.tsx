import React from "react";
import Rating from "@material-ui/lab/Rating";
import { makeStyles } from "@material-ui/core/styles";
import Card from "@material-ui/core/Card";
import CardActionArea from "@material-ui/core/CardActionArea";
import CardContent from "@material-ui/core/CardContent";
import Typography from "@material-ui/core/Typography";
import { FavItem } from "../store/fav-list/types";

const useStyles = makeStyles({
  card: {
    borderRadius: 0
  }
});

interface FavouriteItemProps {
  item: FavItem;
  updateRating: (favItem: FavItem, rating: number) => void;
}

const FavouriteItem: React.FC<FavouriteItemProps> = ({
  item,
  updateRating
}) => {
  const classes = useStyles();

  const onChangeRating = (event: any, rating: number | null) => {
    if (rating) {
      const favItem = { ...item };
      updateRating(favItem, rating);
    }
  };

  return item.rating > 0 && item.rating <= 5 ? (
    <Card className={classes.card}>
      <CardActionArea>
        <CardContent>
          <Typography
            data-testid="name"
            gutterBottom
            variant="h5"
            component="h2"
          >
            {item.name}
          </Typography>
          <Typography
            data-testid="author"
            gutterBottom
            variant="body2"
            color="textSecondary"
            component="p"
          >
            {item.author}
          </Typography>
          <Typography component="p">
            <Rating
              name={`rating${item.id}`}
              size="large"
              value={item.rating}
              precision={0.5}
              onChange={onChangeRating}
              max={5}
            />
          </Typography>
        </CardContent>
      </CardActionArea>
    </Card>
  ) : (
    <span data-testid="invalidRating">Invalid Rating</span>
  );
};
export default FavouriteItem;
