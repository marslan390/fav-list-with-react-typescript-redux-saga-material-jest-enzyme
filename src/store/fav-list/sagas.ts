import { GET_FAV_LIST } from "./types";
import { getApi } from "../../utils/api";
import { call, put, takeLatest } from "redux-saga/effects";
import { receiveFavList, receiveFavListError } from "./actions";

function* _getFavList() {
  const { data, status } = yield call(getApi, {
    url: "/data/fav-list.json"
  });
  status === 200
    ? yield put(receiveFavList(data))
    : yield put(receiveFavListError({}));
}

export function* getFavListSaga() {
  yield takeLatest(GET_FAV_LIST, _getFavList);
}
