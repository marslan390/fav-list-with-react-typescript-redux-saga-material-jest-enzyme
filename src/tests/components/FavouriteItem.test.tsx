import React from "react";
import {shallow} from "enzyme";
import FavouriteItem from "../../components/FavouriteItem";
import { FavItem } from "../../store/fav-list/types";
import Rating from "@material-ui/lab/Rating";

describe("<FavouriteItem/>", () => {
  let props: any;

  beforeEach(() => {
    props = {
      item: {name: "Book", author: "Arslan", rating: 5},
      updateRating: (favItem: FavItem, rating: number) => {}
  }
  });

  test("renders correctly", () => {
    const wrapper = shallow(<FavouriteItem {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("renders fav item heading", () => {
    const wrapper = shallow(<FavouriteItem {...props} />);
    expect(wrapper.find({ "data-testid": "name" }).text()).toEqual("Book");
  });

  test("renders fav item author", () => {
    const wrapper = shallow(<FavouriteItem {...props} />);
    expect(wrapper.find({ "data-testid": "author" }).text()).toEqual("Arslan");
  });

  test("renders Rating component under FavouriteItem", () => {
    const wrapper = shallow(<FavouriteItem {...props} />);
    expect(wrapper.find(Rating)).toHaveLength(1);
  });

  test("renders correct rating value", () => {
    const wrapper = shallow(<FavouriteItem {...props} />);
    expect(wrapper.find(Rating).prop("value")).toEqual(5);
  });

  test("tests invalid rating", () => {
    props = { ...props, item: {...props.item, rating: -1} };
    let wrapper = shallow(<FavouriteItem {...props} />);
    expect(wrapper.find({ "data-testid": "invalidRating" }).text()).toEqual(
      "Invalid Rating"
    );

    props = { ...props, item: {...props.item, rating: 6} };
    wrapper = shallow(<FavouriteItem  {...props} />);
    expect(wrapper.find({ "data-testid": "invalidRating" }).text()).toEqual(
      "Invalid Rating"
    );
  });
});
