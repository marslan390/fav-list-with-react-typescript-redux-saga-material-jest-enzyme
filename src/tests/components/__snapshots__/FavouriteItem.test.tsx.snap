// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavouriteItem/> renders correctly 1`] = `ShallowWrapper {}`;
